src/contract_code.rs

```diff
@@ -0,0 +1,201 @@
+use crate::error::DecodingErr;
+use crate::rlp::{FromRlpItem, RlpItem, ToRlpItem};
+use crate::Bytes;
+
+// TODO: this should come from another module which has not been rewritten yet
+/// Identifier tag of serialized contract code
+const CODE_TAG: u8 = 70;
+
+/// Contract format version.
+const VSN: u8 = 3;
+
+/// FATE contract code with metadata
+#[derive(Debug, PartialEq)]
+pub struct Code {
+    /// Byte code of the contract.
+    pub byte_code: Bytes,
+    /// Whether the contract can receive tokens through a `spend` transaction.
+    pub payable: bool,
+    /// Hash of the source code in the original smart contract language. Note that verification of
+    /// this field is not imposed by the æternity protocol, thus its validity has to always be
+    /// checked before the contract is used.
+    pub source_hash: Bytes,
+    /// Version of the compiler of the original smart contract language. Note that verification of
+    /// this field is not imposed by the æternity protocol, thus its validity has to always be
+    /// checked before the contract is used.
+    pub compiler_version: Bytes,
+}
+
+impl Code {
+    /// Serialize FATE code as a byte-encoded RLP object.
+    pub fn serialize(self: &Code) -> Bytes {
+        self.serialize_rlp()
+    }
+
+    /// Deserializes a byte-encoded RLP object into FATE code.
+    pub fn deserialize(bytes: &[u8]) -> Result<Code, DecodingErr> {
+        FromRlpItem::deserialize_rlp(bytes)
+    }
+}
+
+impl ToRlpItem for Code {
+    fn to_rlp_item(&self) -> RlpItem {
+        let fields = vec![
+            // Tag
+            CODE_TAG.to_rlp_item(), // TODO: should not be hardcoded
+            // Contract version
+            VSN.to_rlp_item(), // TODO: should this be hardcoded?
+            // Source hash
+            RlpItem::ByteArray(self.source_hash.to_vec()),
+            // Type info (AEVM residue, has to be empty)
+            RlpItem::List(vec![]),
```

Comment:
If it's not the main goal to produce the same serialized code as the one produced by erlang, then why do you have to keep this as an empty list? This will be serialized as a single byte `192`. Why not just remove it?

Reply:
It works one way for now:

- All RLP we produce has to be accepted by the chain
- Not every RLP given to us by the chain has to be processed by us.

The only ones we do not process are AEVM contracts and FATE contracts polluted with useless AEVM bloat.